paper/paper.Rmd

```diff
@@ -0,0 +1,221 @@
+
+# Features
+
+As of this writing, `{lintr}` offers `r length(all_linters())` linters.
+
+```{r all_linters}
+library(lintr)
+
+length(all_linters())
+```
+
+Naturally, we can't discuss all of them here. To see details about all available linters, we encourage readers to see <https://lintr.r-lib.org/dev/reference/index.html#individual-linters>.
+
+We will showcase one linter for each kind of common problem found in R code.
+
+- **Best practices**
+
+`{lintr}` offers linters that can detect problematic antipatterns and suggest alternatives that follow best practices.
+
+For example, expressions like `ifelse(x, TRUE, FALSE)` and `ifelse(x, FALSE, TRUE)` are redundant; just `x` or `!x` suffice in R code where logical vectors are a core data structure. The `redundant_ifelse_linter()` linter detects such discouraged usages.
+
+```{r redundant_ifelse_linter}
+lint(
+  text = "ifelse(x >= 2.5, TRUE, FALSE)",
+  linters = redundant_ifelse_linter()
+)
+```
+
+- **Efficiency**
+
+Sometimes the users might not be aware of a more efficient way offered by R for carrying out a computation. `{lintr}` offers linters to improve code efficiency by avoiding common inefficient patterns.
+
+For example, the `any_is_na_linter()` linter detects usages of `any(is.na(x))` and suggests `anyNA(x)` as a more efficient alternative to detect presence of *any* missing values.
+
+```{r any_is_na_linter_with_lint}
+lint(
+  text = "any(is.na(x), na.rm = TRUE)",
+  linters = any_is_na_linter()
+)
+```
+
+`anyNA()` in R is more efficient than `any(is.na())` because it stops execution once a missing value is found, while `is.na()` evaluates the entire vector.
+
+```{r any_is_na_linter_without_lint}
+lint(
+  text = "anyNA(x)",
+  linters = any_is_na_linter()
+)
+```
+
+- **Readability**
+
+Coders spend significantly more time reading than writing code [@mcconnell2004code]. Thus, writing readable code makes the code more maintainable and reduces the possibility of introducing bugs stemming from a poor understanding of the code.
+
+`{lintr}` provides a number of linters that suggest more readable alternatives. For example, `comparison_negation_linter()` blocks usages like `!(x == y)` where a direct relational operator is appropriate.
+
+```{r comparison_negation_linter_with_lint}
+lint(
+  text = "!x == 2",
+  linters = comparison_negation_linter()
+)
+```
+
+The more readable alternative here uses `!=`:
```

Comment:
Blanking on the best phrasing here, but I think it's worth calling out that `!x==2` is even more unreadable than `!(x==2)`.

```suggestion
Note also the complicated operator precedence. The more readable alternative here uses `!=`:
```